sql/sql_table.cc

```diff
@@ -4445,25 +4447,43 @@ static bool vers_prepare_keys(THD *thd, HA_CREATE_INFO *create_info,
     if (key->type != Key::PRIMARY && key->type != Key::UNIQUE)
       continue;
 
-    Key_part_spec *key_part= NULL;
-    List_iterator<Key_part_spec> part_it(key->columns);
-    while ((key_part=part_it++))
+    if (create_info->versioned())
     {
-      if (!my_strcasecmp(system_charset_info,
-                         row_start_field,
-                         key_part->field_name.str) ||
-
-          !my_strcasecmp(system_charset_info,
-                         row_end_field,
-                         key_part->field_name.str))
-        break;
+      Key_part_spec *key_part=NULL;
+      List_iterator<Key_part_spec> part_it(key->columns);
+      while ((key_part=part_it++))
+      {
+        if (row_start_field.streq(key_part->field_name) ||
+            row_end_field.streq(key_part->field_name))
+          break;
+      }
+      if (key_part)
+        key->columns.push_back(new Key_part_spec(&row_end_field, 0));
     }
-    if (key_part)
-      continue; // Key already contains Sys_start or Sys_end
 
-    Key_part_spec *key_part_sys_end_col=
-        new (thd->mem_root) Key_part_spec(&create_info->vers_info.as_row.end, 0);
-    key->columns.push_back(key_part_sys_end_col);
+    if (key->without_overlaps)
+    {
+      DBUG_ASSERT(create_info->period_info.is_set());
```

Comment:
## Bug 1: assertion on CREATE partitioned table
### Reproduce
```sql
create or replace table t1 (
  x int, s date, e date,
  primary key(x, p without overlaps))
partition by hash (x) partitions 2;
```
### Result
```c++
#3  0x00007ffff6d16142 in __GI___assert_fail (assertion=0x14c57b9 "create_info->period_info.is_set()", file=0x14c3c99 "/home/midenok/src/mariadb/review/src/sql/sql_table.cc", line=4466, function=0x14c5767 "bool append_system_key_parts(THD *, HA_CREATE_INFO *, Alter_info *, KEY **, uint)") at assert.c:101
#4  0x0000000000891276 in append_system_key_parts (thd=0x7fff7c000cf8, create_info=0x7ffff42202a0, alter_info=0x7ffff42201d0, key_info=0x7ffff421e280, key_count=0) at /home/midenok/src/mariadb/review/src/sql/sql_table.cc:4466
#5  0x0000000000890bf3 in mysql_create_frm_image (thd=0x7fff7c000cf8, db=0x7fff7c015180, table_name=0x7fff7c015190, create_info=0x7ffff42202a0, alter_info=0x7ffff42201d0, create_table_mode=0, key_info=0x7ffff421e280, key_count=0x7ffff421e27c, frm=0x7ffff421e060) at /home/midenok/src/mariadb/review/src/sql/sql_table.cc:4730
#6  0x00000000008924e7 in create_table_impl (thd=0x7fff7c000cf8, orig_db=0x7fff7c015180, orig_table_name=0x7fff7c015190, db=0x7fff7c015180, table_name=0x7fff7c015190, path=0x7ffff421e070 "./test/t1", options=..., create_info=0x7ffff42202a0, alter_info=0x7ffff42201d0, create_table_mode=0, is_trans=0x7ffff421e3a7, key_info=0x7ffff421e280, key_count=0x7ffff421e27c, frm=0x7ffff421e060) at /home/midenok/src/mariadb/review/src/sql/sql_table.cc:4981
#7  0x00000000008917c8 in mysql_create_table_no_lock (thd=0x7fff7c000cf8, db=0x7fff7c015180, table_name=0x7fff7c015190, create_info=0x7ffff42202a0, alter_info=0x7ffff42201d0, is_trans=0x7ffff421e3a7, create_table_mode=0, table_list=0x7fff7c015168) at /home/midenok/src/mariadb/review/src/sql/sql_table.cc:5110
#8  0x0000000000892d5a in mysql_create_table (thd=0x7fff7c000cf8, create_table=0x7fff7c015168, create_info=0x7ffff42202a0, alter_info=0x7ffff42201d0) at /home/midenok/src/mariadb/review/src/sql/sql_table.cc:5202
#9  0x000000000079e2fb in mysql_execute_command (thd=0x7fff7c000cf8) at /home/midenok/src/mariadb/review/src/sql/sql_parse.cc:4282
#10 0x0000000000797510 in mysql_parse (thd=0x7fff7c000cf8, rawbuf=0x7fff7c014fc0 "create or replace table t1 (\n  x int, s date, e date,\n  primary key(x, p without overlaps))\npartition by hash (x) partitions 2", length=126, parser_state=0x7ffff4221640, is_com_multi=false, is_next_command=false) at /home/midenok/src/mariadb/review/src/sql/sql_parse.cc:8078
```
#### frame 4
```c++
4464        if (key->without_overlaps)
4465        {
4466          DBUG_ASSERT(create_info->period_info.is_set());
```
```
(gdb) p create_info->period_info
$5 = {
  <Sql_alloc> = {<No data fields>}, 
  members of Table_period_info: 
  name = {
    <Lex_cstring> = {
      <st_mysql_const_lex_string> = {
        str = 0x0, 
        length = 0
      }, <No data fields>}, <No data fields>}, 
  period = {
    start = {
      <Lex_cstring> = {
        <st_mysql_const_lex_string> = {
          str = 0x0, 
          length = 0
        }, <No data fields>}, <No data fields>}, 
    end = {
      <Lex_cstring> = {
        <st_mysql_const_lex_string> = {
          str = 0x0, 
          length = 0
        }, <No data fields>}, <No data fields>}
  }, 
  unique_keys = 0
}
```

Reply:
It could be good to have [some cases](https://github.com/tempesta-tech/mariadb/issues/549#issuecomment-430232583) with versioned partitioning.

Reply:
*That's not the bug you are looking for*

You just forgot to add `period for p(s,e)`

Reply:
unofrtunately we partitioning needs additional support implementation, but assertion on missing `period for p(s,e)` is fixed